sending MultiPart in http client Fix #1005

WebFreak001 opened the pull request:
this is also testing the server implementation which made me find #1875

s-ludwig commented on http/vibe/http/common.d:
```diff
@@ -336,13 +337,222 @@ class HTTPStatusException : Exception {
+	import vibe.stream.memory : createMemoryStream;
+
+	InetHeaderMap headers;
+	InputStream content;
```
This needs to be `InterfaceProxy!InputStream` to enable forward compatibility with vibe-core.

s-ludwig commented on http/vibe/http/common.d:
```diff
@@ -336,13 +337,222 @@ class HTTPStatusException : Exception {
+		ret.headers["Content-Type"] = type;
+		if (!type.startsWith("text/"))
+			ret.headers["Content-Transfer-Encoding"] = "binary";
+		ret.content = cast(InputStream) openFile(file, FileMode.read);
```
-> `InterfaceProxy!InputStream(openFile(...))`

s-ludwig commented on http/vibe/http/common.d:
```diff
@@ -336,13 +337,222 @@ class HTTPStatusException : Exception {
+		ret.headers["Content-Type"] = type;
+		if (!type.startsWith("text/"))
+			ret.headers["Content-Transfer-Encoding"] = "binary";
+		ret.content = cast(InputStream) openFile(file, FileMode.read);
```
Same as above.

s-ludwig commented on http/vibe/http/common.d:
```diff
@@ -336,13 +337,222 @@ class HTTPStatusException : Exception {
+		return ret;
+	}
+
+	static MultiPartBodyPart multipleFilesPart(string filename, string content_type, InputStream stream)
```
This should use `isInputStream` to enable proper use when using vibe-core:
```
static MultiPartBodyPart multipleFilesPart(InputStream)(string filename, string content_type, InputStream stream)
    if (isInputStream!InputStream)
{
```

s-ludwig commented on http/vibe/http/common.d:
```diff
@@ -336,13 +337,222 @@ class HTTPStatusException : Exception {
+		multipart.write(boundary, stream);
+		ret.content = createMemoryStream(stream.data, false);
+		return ret;
+	}
```
Looks like the last three overloads can just call the first one to avoid some code duplication. The change to be vibe-core compatible would otherwise add some more duplication for the required `InterfaceProxy` constructor call.

WebFreak001 replied:
I changed it to emplace now which seems to work and changed old code using direct checks on inheritance checks to now access the private typeinfo field of the proxy class because I don't want to create another PR on vibe-core which would take 3 years to merge...

s-ludwig commented on http/vibe/http/common.d:
```diff
@@ -336,13 +337,222 @@ class HTTPStatusException : Exception {
+	}
+}
+
+string randomMultipartBoundary()
```
I'd keep this private for now, since there is no obvious public use for this and it would mean that the exact API isn't critical.

WebFreak001 replied:
this is a public API now and is fairly critical if you want to nest multiparts. Because MultiPart is now in vibe.inet it also doesn't have access to crypto which is now used in this API

s-ludwig commented on http/vibe/http/common.d:
```diff
@@ -336,13 +337,222 @@ class HTTPStatusException : Exception {
 	string debugMessage;
 }
 
+final class MultiPartBodyPart
```
I'd keep this named as `MultiPart` and rename the new `MultiPart` class below to `MultiPartBody`.

The idea for `HTTPClientRequest.writePart` was that it would indeed just write a single part, so that not all parts would have to be constructed up front. For this reason I'd also not expose the new `MultiPart` class below as a public API, but just use the code within to implement the `writePart` method with proper lazy writing of the preamble etc.

WebFreak001 replied:
well the preamble and epilogue are so rarely used that they aren't even documented in the w3c spec, only in some HTTP specs I could find. I think if you write lazily there shouldn't be any support for these

s-ludwig replied:
I definitely wouldn't mind that. If someone should ever actually need them, they can still be added.

WebFreak001 commented on http/vibe/http/common.d:
```diff
@@ -336,13 +337,222 @@ class HTTPStatusException : Exception {
+		multipart.write(boundary, stream);
+		ret.content = createMemoryStream(stream.data, false);
+		return ret;
+	}
```
uh `InterfaceProxy!InputStream(openFile(...))` didn't work because there is no opCall for it, I looked at the definition and it's just aliasing itself and it's just an interface right now so that would break, I changed it to a cast though

s-ludwig replied:
Sorry for the confusion, I looked at a piece of code that actually uses the constructor syntax, but that was actually in a `static if` block that tested for vibe-core.

WebFreak001 replied:
so what should I change it to?

WebFreak001 commented:
@s-ludwig ping

WebFreak001 commented:
@s-ludwig pong

WebFreak001 commented:
@s-ludwig bump... I need this yet again...

WebFreak001 commented:
is there any way to check if the proxy is a subclass of something else? Before it was simply casting and checking if it was null but now it seems to be overly complicated without triggering assert errors...

WebFreak001 commented:
ping

vitalka200 commented:
Any plans on merging this? @s-ludwig

WebFreak001 commented:
bump @s-ludwig 

- improved API
- moved into vibe.inet (because this is a MIME RFC which also applies to mails)
- no longer using my emplace hack which probably broke everything before just because I was too lazy to understand the new vibe-core code, but now instead using proper code
- no internal proxy reading hacks anymore, instead stores the length along with MultiPart
- fixed unittest to actually run
